### PR TITLE
Update snowflake algorithm to support modulus of 2^n

### LINKED ISSUE #3422: version 4.0.0-RC2: key generated by SnowflakeShardingKeyGenerator mod 4 always is 0/1, expected 0/1/2/3
Hi.
I found a fact about the sharding key generated by built-in snowflake algorithm. The result of the key mod 4 always is 0/1, to me, it is wierd. What I expected is, the result should be 0/1/2/3, and the key should evenly distributed to all the sharded tables. I am new to ShardingSphere and the snowflake algorithm. Hope I can get some help here.

**I found out there is  no this problem when keys are generated in a same millisecond.**

Thanks.

`public final class PreciseModuloShardingTableAlgorithm implements PreciseShardingAlgorithm<Long> {

    @Override
    public String doSharding(final Collection<String> tableNames, final PreciseShardingValue<Long> shardingValue) {

        String tableName = (String) tableNames.toArray()[0];

        if(tableName.startsWith("table_a")) {
            for (String each : tableNames) {
                if (each.endsWith(shardingValue.getValue() % 4 + "")) {
                    return each;
                }
            }
        }

        if (tableName.startsWith("table_b")) {
            for (String each : tableNames) {
                if (each.endsWith(shardingValue.getValue() % 2 + "")) {
                    return each;
                }
            }
        }

        throw new UnsupportedOperationException();
    }
}`

In database config class:
`    DataSource getShardingDataSource() throws SQLException {
        ShardingRuleConfiguration shardingRuleConfig = new ShardingRuleConfiguration();

        shardingRuleConfig.getTableRuleConfigs().add(new TableRuleConfiguration("table_a", "ds.table_a_$->{0..3}"));
        shardingRuleConfig.getBindingTableGroups().add("table_a");

        shardingRuleConfig.getTableRuleConfigs().add(new TableRuleConfiguration("table_b", "ds.admin_behaviour_track_$->{0..2}"));
        shardingRuleConfig.getBindingTableGroups().add("table_b");

        shardingRuleConfig.setDefaultKeyGeneratorConfig(new KeyGeneratorConfiguration("SNOWFLAKE", "id"));

        shardingRuleConfig.setDefaultTableShardingStrategyConfig(new StandardShardingStrategyConfiguration("id", new PreciseModuloShardingTableAlgorithm()));

        Properties properties = new Properties();
        properties.setProperty(ShardingPropertiesConstant.SQL_SHOW.getKey(), "true");

        return ShardingDataSourceFactory.createDataSource(createDataSourceMap(), shardingRuleConfig, properties);
    }`

### PR DESCRIPTION
**Pull request by geomonlin**
Fixes #3422 .

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

**Comment by geomonlin** [comment]
> Please add test case

I've added test cases, but I‘m not sure if I haven't noticed anything

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java [hunk 1]
```diff
@@ -158,6 +167,27 @@ public void assertSetWorkerIdFailureWhenTooMuch() {
...
+    @SneakyThrows
+    public void assertSetMaxVibrationOffsetFailureWhenNPowerOf2() {
+        SnowflakeShardingKeyGenerator keyGenerator = new SnowflakeShardingKeyGenerator();
+        Properties properties = new Properties();
+        properties.setProperty("max.vibration.offset", String.valueOf(2));
+        keyGenerator.setProperties(properties);
+        Field props = keyGenerator.getClass().getDeclaredField("properties");
+        props.setAccessible(true);
+        assertThat(((Properties) props.get(keyGenerator)).get("max.vibration.offset"), is((Object) "2"));
+    }
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java]: Should assert result of key generated for last digital

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java [hunk 2]
```diff
@@ -181,4 +211,16 @@ public void assertSetMaxTolerateTimeDifferenceMilliseconds() {
         props.setAccessible(true);
         assertThat(((Properties) props.get(keyGenerator)).get("max.tolerate.time.difference.milliseconds"), is((Object) "1"));
     }
+    
+    @Test
+    @SneakyThrows
+    public void assertSetMaxVibrationOffsetSuccess() {
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java]: Should assert result of key generated for last digital

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java [hunk 3]
```diff
@@ -158,6 +167,32 @@ public void assertSetWorkerIdFailureWhenTooMuch() {
...
+        SnowflakeShardingKeyGenerator keyGenerator = new SnowflakeShardingKeyGenerator();
+        Properties properties = new Properties();
+        properties.setProperty("max.vibration.offset", String.valueOf(4096));
+        keyGenerator.setProperties(properties);
+        keyGenerator.generateKey();
+    }
+    
+    @Test
+    @SneakyThrows
+    public void assertSetMaxVibrationOffsetFailureWhenNPowerOf2() {
```
Comment (KomachiSion) [inline sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java]: I read the test method name seems like I will get exception or set offset failure when the values if power of 2.
But your test case seem to assert set successfully.
So I think the test has some error.

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java [hunk 4]
```diff
@@ -76,6 +76,21 @@ public void assertGenerateKeyWithSingleThread() {
...
+        keyGenerator.setProperties(properties);
+        SnowflakeShardingKeyGenerator.setTimeService(new FixedTimeService(1));
+        List<Comparable<?>> expected = Arrays.<Comparable<?>>asList(1L, 4194306L, 4194307L, 8388611L, 8388612L, 12582912L, 12582913L, 16777217L, 16777218L, 20971522L);
+        List<Comparable<?>> actual = new ArrayList<>();
+        for (int i = 0; i < DEFAULT_KEY_AMOUNT; i++) {
+            actual.add(keyGenerator.generateKey());
+        }
+        assertThat(actual, is(expected));
+    }
+
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java]: The test case is still not clear, we just case about the last digital of generated keys, right?
The first case, we can assert the last digital of generated keys for each milliseconds as `0`, `1`, `2`, `3` and `0` again.
The second case, we can assert the last digital of generated keys in same milliseconds as accumulate over the Vibration offset.

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java [hunk 5]
```diff
@@ -76,6 +76,21 @@ public void assertGenerateKeyWithSingleThread() {
         assertThat(actual, is(expected));
     }
     
+    @Test
+    public void assertGenerateKeyWithMaxVibrationOffsetBe1LessThanPowerOf2() {
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java]: What is the mean of the test case name: `assertGenerateKeyWithMaxVibrationOffsetBe1LessThanPowerOf2`?

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java [hunk 6]
```diff
@@ -149,6 +164,24 @@ public void assertSetWorkerIdFailureWhenNegative() {
...
+    public void assertSetMaxVibrationOffsetFailureWhenNegative() {
+        SnowflakeShardingKeyGenerator keyGenerator = new SnowflakeShardingKeyGenerator();
+        Properties properties = new Properties();
+        properties.setProperty("max.vibration.offset", String.valueOf(-1));
+        keyGenerator.setProperties(properties);
+        keyGenerator.generateKey();
+    }
+    
+    @Test(expected = IllegalArgumentException.class)
+    public void assertSetMaxVibrationOffsetFailureWhenZero() {
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java]: zero of vibration offset should be fine, which means every milliseconds they are start from zero, is it?

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java [hunk 7]
```diff
@@ -149,6 +164,24 @@ public void assertSetWorkerIdFailureWhenNegative() {
...
+        SnowflakeShardingKeyGenerator keyGenerator = new SnowflakeShardingKeyGenerator();
+        Properties properties = new Properties();
+        properties.setProperty("max.vibration.offset", String.valueOf(0));
+        keyGenerator.setProperties(properties);
+        keyGenerator.generateKey();
+    }
+    
     @Test(expected = IllegalArgumentException.class)
     public void assertSetWorkerIdFailureWhenTooMuch() {
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java]: rename `assertSetWorkerIdFailureWhenTooMuch` to `assertSetWorkerIdFailureWhenOutOfRange` is better

File: sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java [hunk 8]
```diff
@@ -181,4 +223,16 @@ public void assertSetMaxTolerateTimeDifferenceMilliseconds() {
         props.setAccessible(true);
         assertThat(((Properties) props.get(keyGenerator)).get("max.tolerate.time.difference.milliseconds"), is((Object) "1"));
     }
+    
+    @Test
+    @SneakyThrows
+    public void assertSetMaxVibrationOffsetSuccess() {
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/test/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGeneratorTest.java]: Do we need this test case? Does it duplicate with `assertGenerateKeyWithMaxVibrationOffsetBe1LessThanPowerOf2`?
